### PR TITLE
[KGE] Kg loader

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Add support for user defined dataset for DGL KGE package

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: apps/kg/dataloader/KGDataset.py [hunk 1]
```diff
@@ -135,7 +135,13 @@ class KGDatasetFB15k237(KGDataset):
     '''
     def __init__(self, path, name='FB15k-237', read_triple=True, only_train=False):
         self.name = name
-        url = 'https://s3.us-east-2.amazonaws.com/dgl.ai/dataset/{}.zip'.format(name)
+        url = '
```
Comment (VoVAllen) [inline apps/kg/dataloader/KGDataset.py]: What's happening here?

File: apps/kg/README.md [hunk 2]
```diff
@@ -56,6 +56,21 @@ DGL-KE provides five knowledge graphs:
...
+  - *relations* stores the mapping between relation name relation Id
+  - *train* stores the triples in the training set. In format [src_name, rel_name, dst_name]
+  - *valid* stores the triples in the validation set. In format [src_name, rel_name, dst_name]
+  - *test* stores the triples in the test set. In format [src_name, rel_name, dst_name]
+- udd, user defined dataset format. it should contains five files:
+  - *entities* stores the mapping between entity name and entity Id
+  - *relations* stores the mapping between relation name relation Id
+  - *train* stores the triples in the training set. In format [src_id, rel_id, dst_id]
+  - *valid* stores the triples in the validation set. In format [src_id, rel_id, dst_id]
+  - *test* stores the triples in the test set. In format [src_id, rel_id, dst_id]
```
Comment (zheng-da) [inline apps/kg/README.md]: the difference is that one provides triplets of entity names and relation names. the other provides triplets of ids?

File: apps/kg/dataloader/KGDataset.py [hunk 3]
```diff
@@ -23,61 +23,62 @@ def _download_and_extract(url, path, filename):
...
         heads = []
         tails = []
         rels = []
-        with open(os.path.join(path, '{}.txt'.format(mode))) as f:
+        with open(path) as f:
+            if skip_first_line:
+                _ = f.readline()
             for line in f:
                 h, r, t = line.strip().split('\t')
                 heads.append(self.entity2id[h])
```
Comment (zheng-da) [inline apps/kg/dataloader/KGDataset.py]: what if the entities and relations are already integers?

File: apps/kg/eval.py [hunk 4]
```diff
@@ -29,8 +29,11 @@ def __init__(self):
                           help='root path of all dataset')
         self.add_argument('--dataset', type=str, default='FB15k',
                           help='dataset name, under data_path')
-        self.add_argument('--format', type=str, default='1',
+        self.add_argument('--format', type=str, default='built_in',
```
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/eval.py]: Done in the new push

**Comment by classicsong** [comment]
> we also need a preprocessing script to prepare users' data in DGLGraph format to accelerate DGL loading in the future.

What is DGLGraph format?

File: apps/kg/README.md [hunk 5]
```diff
@@ -129,24 +129,20 @@ when given (?, rel, tail).
...
+DGL-KE supports two knowledge graph input formats for user defined dataset
 
-Format 1:
-
-- entities.dict contains pairs of (entity Id, entity name). The number of rows is the number of entities (nodes).
-- relations.dict contains pairs of (relation Id, relation name). The number of rows is the number of relations.
-- train.txt stores edges in the training set. They are stored as triples of (head, rel, tail).
-- valid.txt stores edges in the validation set. They are stored as triples of (head, rel, tail).
-- test.txt stores edges in the test set. They are stored as triples of (head, rel, tail).
+- raw_udd_[h|r|t], raw user defined dataset. In this format, user only need to provide triples and let the dataloader generate and manipulate the id mapping. The dataloader will generate two files: entities.tsv for entity id mapping and relations.tsv for relation id mapping. The order of head, relation and tail entities are described in [h|r|t], for example, raw_udd_drs means the triples are stored in the order of tail, relation and head. It should contains three files:
```
Comment (zheng-da) [inline apps/kg/README.md]: what does `raw_udd_drs` mean?

File: apps/kg/README.md [hunk 6]
```diff
@@ -129,24 +129,20 @@ when given (?, rel, tail).
...
+  - *test* stores the triples in the test set. In format of a triple, e.g., [src_name, rel_name, dst_name] and should follow the order specified in [h|r|t]
 
 Format 2:
-
-- entity2id.txt contains pairs of (entity name, entity Id). The number of rows is the number of entities (nodes).
-- relation2id.txt contains pairs of (relation name, relation Id). The number of rows is the number of relations.
-- train.txt stores edges in the training set. They are stored as triples of (head, tail, rel).
-- valid.txt stores edges in the validation set. They are stored as a triple of (head, tail, rel).
-- test.txt stores edges in the test set. They are stored as a triple of (head, tail, rel).
+- udd_[h|r|t], user defined dataset. In this format, user should provide the id mapping for entities and relations. The order of head, relation and tail entities are described in [h|r|t], for example, raw_udd_drs means the triples are stored in the order of tail, relation and head. It should contains five files:
```
Comment (zheng-da) [inline apps/kg/README.md]: why is it `raw_udd_drs` again?

File: apps/kg/dataloader/KGDataset.py [hunk 7]
```diff
@@ -150,27 +291,201 @@ def read_triple(self, path, mode, skip_first_line=False):
...
+            dataset = KGDatasetFB15k237(data_path, part_name, read_triple=True, only_train=True)
+        elif data_name == 'wn18':
+            dataset = KGDatasetWN18(data_path, part_name, read_triple=True, only_train=True)
+        elif data_name == 'wn18rr':
+            dataset = KGDatasetWN18rr(data_path, part_name, read_triple=True, only_train=True)
+        else:
+            assert False, "Unknown dataset {}".format(data_name)
+    elif format_str == 'raw_udd':
+        # user defined dataset
+        assert False, "When using partitioned dataset, we assume dataset will not be raw"
```
Comment (zheng-da) [inline apps/kg/dataloader/KGDataset.py]: why is it not available yet?

File: apps/kg/dataloader/KGDataset.py [hunk 8]
```diff
@@ -150,27 +291,201 @@ def read_triple(self, path, mode, skip_first_line=False):
...
+        self.name = name
+        for f in files:
+            assert os.path.exists(os.path.join(path, f)), \
+                'File {} now exist in {}'.format(f, path)
+
+        assert len(format) == 3
+        format = _parse_srd_format(format)
+        self.load_entity_relation(path, files, format)
+
+        if len(files) == 1:
```
Comment (zheng-da) [inline apps/kg/dataloader/KGDataset.py]: I suppose it means the training data is provided not not validation data and test data.

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply apps/kg/eval.py]: how does a user provide the definition of a triple?

File: apps/kg/train.py [hunk 9]
```diff
@@ -30,8 +30,10 @@ def __init__(self):
                           help='root path of all dataset')
         self.add_argument('--dataset', type=str, default='FB15k',
                           help='dataset name, under data_path')
-        self.add_argument('--format', type=str, default='1',
+        self.add_argument('--format', type=str, default='built_in',
```
Comment (zheng-da) [inline apps/kg/train.py]: do we support the same format as eval.py?

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/eval.py]: Here we assume the file is a tsv with three columns

File: apps/kg/README.md (hunk 5 above)
Comment (classicsong) [reply apps/kg/README.md]: It should be changed to raw_udd_trh.

File: apps/kg/dataloader/KGDataset.py (hunk 7 above)
Comment (classicsong) [reply apps/kg/dataloader/KGDataset.py]: For distributed training, we should ask user to provide processed dataset. As each partition may not have the whole entity nodes, generating id mapping is dangerous.

File: apps/kg/train.py (hunk 9 above)
Comment (classicsong) [reply apps/kg/train.py]: Same for eval.py

File: apps/kg/eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply apps/kg/eval.py]: And for raw_udd and udd we ask use to provide htr hint in a way of raw_udd_htr.

File: apps/kg/README.md (hunk 2 above)
Comment (classicsong) [reply apps/kg/README.md]: Yes, now we have udd_raw with only triplets as input and udd with entity map/relation map and triplets as input

**Pull request merged**